# Mini Project 1 — Analysis Notebook

**Your name: Nicolas Rodriguez**  
**Dataset: Spotify API**  
**Date: May 11, 2026**  



## Setup Cell
Run the cell below first. 

In [2]:
# Setup — run this cell first
!pip install jupyter plotly kaleido pandas

import pandas as pd
import plotly.express as px

print("Setup complete.")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 914.9/914.9 kB 16.2 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 19.8 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.5/12.5 MB 21.1 MB/s  0:00:00eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 17.8 MB/s  0:00:00eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 17.9 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.6/14.6 MB 14.3 MB/s  0:00:01 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56/56 [jupyter]4/56 [notebook]b]ver]
Setup complete.


---

## Section 1 — Overview

Before writing any code, fill in this section. A good Overview tells anyone reading your notebook — including a future employer — what the analysis is about before they see a single chart.

**Dataset:** The data I have chosen for this project comes from Spotify's API. It includes the data of my recent listening history. The API URL is the following *https://api.spotify.com/v1* 

**Why this dataset:** This dataset would allow me to develop a basic version of the yearly Spotify Wrapped, I wished to recreate the data analysis and visualization behind a feature loved by millions of users. 

**Three analytical questions:**

1. *How much overlap exists between my top artists across a short term (4 weeks) medium term (6 months) and long term (since having the account), what does this reveal about wether my taste is stable or shifting?* 
2. *What is the release year distribution of my top tracks across time ranges, do I gravitate to an older catalog or newer releases?* 
3. *From my 50 most recently played tracks, what proportion of listens come from albums vs playlist vs other contexts, which artists do I listen to most in album mode vs playlist mode?*

**What a practitioner would do with these findings:** In my opinion, this data could be used by many people in product but personally as a product manager I would use this to illustrate how simple ideas can create value to a brand. 

---

## Section 2 — Data Profile

Load your dataset and get a basic picture of what's in it. Answer these questions in a markdown cell below your code:

- How many rows and columns does your dataset have?
- What does each column represent?
- Are there any obvious data quality issues (missing values, unexpected types, inconsistent formatting)?
- Which column or columns will your analysis focus on, and why?

In [3]:
# Load your dataset
# Example (app review dataset from class):
# df = pd.read_csv('app_reviews_demo.csv')

import pandas as pd

df = pd.read_csv('Spotify NR data.csv')  # replace with your filename

print(df.shape)
df.head()

(110, 13)


,source,time_range_key,rank,artist,name,album,release_year,duration_sec,explicit,time_range,track,played_at,context
0,top_artists,short_term,1.0,Pink Floyd,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,top_artists,short_term,2.0,Yandel,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,top_artists,short_term,3.0,Feid,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,top_artists,short_term,4.0,Radiohead,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,top_artists,short_term,5.0,Omar Courtz,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
# Check column types and missing values
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 110 entries, 0 to 109
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   source          110 non-null    str    
 1   time_range_key  30 non-null     str    
 2   rank            30 non-null     float64
 3   artist          110 non-null    str    
 4   name            30 non-null     str    
 5   album           30 non-null     str    
 6   release_year    80 non-null     float64
 7   duration_sec    30 non-null     float64
 8   explicit        30 non-null     object 
 9   time_range      30 non-null     str    
 10  track           50 non-null     str    
 11  played_at       50 non-null     str    
 12  context         50 non-null     str    
dtypes: float64(3), object(1), str(9)
memory usage: 11.3+ KB


In [5]:
# Summary statistics for numeric columns
df.describe()

,rank,release_year,duration_sec
count,30.000000,80.000000,30.000000
mean,5.500000,2017.375000,277.200000
std,2.921384,13.300209,194.099583
min,1.000000,1975.000000,104.000000
25%,3.000000,2017.000000,169.250000
50%,5.500000,2021.000000,213.000000
75%,8.000000,2025.000000,280.500000
max,10.000000,2026.000000,813.000000


In [6]:
# Missing values per column
df.isnull().sum()

source             0
time_range_key    80
rank              80
artist             0
name              80
album             80
release_year      30
duration_sec      80
explicit          80
time_range        80
track             60
played_at         60
context           60
dtype: int64

**Data profile notes:**  
**Q1: How many rows and columns does your dataset have?**  
**A:** The dataset has **110 rows and 13 columns** (`df.shape`).  

**Q2: What does each column represent?**  
**A:** `source` labels the endpoint (`top_artists`, `top_tracks`, `recently_played`). Ranking-window fields are `time_range_key`, `time_range`, and `rank`; track metadata fields are `artist`, `name`, `album`, `release_year`, `duration_sec`, and `explicit`; recent-listening event fields are `track`, `played_at`, and `context`.  

**Q3: Are there any obvious data quality issues (missing values, unexpected types, inconsistent formatting)?**  
**A:** The largest issue is missing values, but they are mostly **structural** because rows from different endpoints only populate the columns relevant to that endpoint (for example, `top_artists` rows lack track-detail columns, while `recently_played` rows lack ranking columns). There are also type-cleaning needs: `rank` and `release_year` load as float when nulls are present, and `played_at` loads as string before datetime parsing. It's also worth noting a recent Spotify update which deprecated endpoints reduced some accessible fields, but the available data is still usable for this project, so I continued with the current scope.  

**Q4: Which column or columns will your analysis focus on, and why?**  
**A:** For Q1 (artist overlap/stability), I use `artist`, `time_range_key`, and `rank`; for Q2 (release-year patterns), I use `name`, `artist`, `release_year`, and `time_range`; for Q3 (recent listening behavior), I use `track`, `artist`, `played_at`, and `context`. These columns directly map to each research question.

---

## Section 3 — Analysis

Answer your three research questions using pandas. Each question should have:

1. A markdown cell stating the question
2. A code cell with the analysis
3. A markdown cell with your interpretation — what does the result mean?

*(I have cleaned the data in the following scripts, while also running the analysis scripts. I chose to do this since the data was fairly well structured and a small amount so a quick clean while running the analysis worked perfect)*

**Question 1:** How do my top artist rankings change across short-, medium-, and long-term listening windows?

In [7]:

# Build top-10 artist lists by time range, then compare overlap across windows.

from IPython.display import display

required_cols = {"source", "artist", "time_range_key", "rank"}
if "df" not in globals() or not required_cols.issubset(df.columns):
    raise NameError("Run the dataset loading cell first so df is available.")

artists_source = (
    df.loc[df["source"] == "top_artists", ["artist", "time_range_key", "rank"]]
    # Remove rows missing fields needed to rank artists by window.
    .dropna(subset=["artist", "time_range_key", "rank"])
    .copy()
)
# Cast rank to integer so sorting/comparison uses true rank values.
artists_source["rank"] = artists_source["rank"].astype(int)

# Keep this dataframe for visualization in Section 4.
time_order = ["short_term", "medium_term", "long_term"]
artists_rank_df = artists_source.copy()
artists_rank_df["time_range_key"] = pd.Categorical(
    artists_rank_df["time_range_key"], categories=time_order, ordered=True
)
artists_rank_df = artists_rank_df.sort_values(["artist", "time_range_key"])

# Script 5 behavior: collect top artists for each time window.
time_ranges = {
    "short_term": "Last 4 Weeks",
    "medium_term": "Last 6 Months",
    "long_term": "All Time",
}

top_artists = {}
for range_key in time_ranges:
    artists = (
        artists_source.loc[artists_source["time_range_key"] == range_key]
        .sort_values("rank")["artist"]
        .head(10)
        .tolist()
    )
    top_artists[range_key] = artists

# Table 1: Top artists by rank across time windows.
rank_table = (
    artists_rank_df
    .pivot_table(index="rank", columns="time_range_key", values="artist", aggfunc="first")
    .rename(columns=time_ranges)
    .sort_index()
)
rank_table.index.name = "Rank"

# Script 6 behavior: overlap and listener profile.
short = set(top_artists["short_term"])
medium = set(top_artists["medium_term"])
long = set(top_artists["long_term"])

short_medium = short & medium
medium_long = medium & long
all_three = short & medium & long
only_short = short - medium - long

if len(all_three) >= 6:
    profile = "Loyalist - you return to the same artists consistently"
elif len(all_three) >= 3:
    profile = "Balanced - mix of core favorites and new discoveries"
else:
    profile = "Explorer - your taste shifts significantly over time"

# Table 2: Overlap summary and listener profile.
overlap_summary = pd.DataFrame(
    {
        "Metric": [
            "Overlap: short vs medium",
            "Overlap: medium vs long",
            "Overlap: all three",
            "Listener profile",
        ],
        "Value": [
            f"{len(short_medium)}/10",
            f"{len(medium_long)}/10",
            f"{len(all_three)}/10",
            profile,
        ],
    }
)

# Table 3: Which artists are core vs newly appearing.
artist_grouping = pd.DataFrame(
    {
        "Core artists (all three windows)": pd.Series(sorted(all_three)),
        "New discoveries (short-term only)": pd.Series(sorted(only_short)),
    }
)

print("Top artists by rank across listening windows")
display(rank_table)

print("\nOverlap summary")
display(overlap_summary)

print("\nCore artists vs new discoveries")
display(artist_grouping)


Top artists by rank across listening windows


time_range_key,Last 4 Weeks,Last 6 Months,All Time
Rank,,,
1,Pink Floyd,Pink Floyd,Pink Floyd
2,Yandel,Feid,Feid
3,Feid,Radiohead,Radiohead
4,Radiohead,Omar Courtz,Mac DeMarco
5,Omar Courtz,Milo j,Mac Miller
6,Olivia Dean,Mac DeMarco,The Doors
7,Trueno,Olivia Dean,Bad Bunny
8,Harry Styles,Brent Faiyaz,Brent Faiyaz
9,Lenny Kravitz,Yandel,Jeff Buckley



Overlap summary


,Metric,Value
0,Overlap: short vs medium,6/10
1,Overlap: medium vs long,5/10
2,Overlap: all three,3/10
3,Listener profile,Balanced - mix of core favorites and new disco...



Core artists vs new discoveries


,Core artists (all three windows),New discoveries (short-term only)
0,Feid,Harry Styles
1,Pink Floyd,Lenny Kravitz
2,Radiohead,Nsqk
3,NaN,Trueno


**Interpretation:**  
Across short-, medium-, and long-term windows, my top-artist profile is mostly stable with selective short-term shifts: `rank_table` shows recurring leaders (for example, Pink Floyd, Feid, and Radiohead), `overlap_summary` confirms meaningful overlap across windows, and `artist_grouping` isolates which names are core versus newly appearing. Together, these outputs answer the question by showing that my taste has a consistent backbone, while shorter windows capture recent rotations rather than a complete change in preference.

**Question 2:** *What is the release year distribution of my top tracks across time ranges, do I gravitate to an older catalog or newer releases?*

In [8]:

# Analyze release-year distribution of top tracks across time ranges.

from IPython.display import display
import plotly.express as px

required_cols = {"source", "name", "artist", "album", "release_year", "time_range"}
if "df" not in globals() or not required_cols.issubset(df.columns):
    raise NameError("Run the dataset loading cell first so df is available.")

# Pull only top-track records from the exported Spotify dataset.
df_tracks = (
    df.loc[df["source"] == "top_tracks", ["name", "artist", "album", "release_year", "duration_sec", "explicit", "time_range"]]
    # Keep only complete records needed for release-year analysis.
    .dropna(subset=["name", "artist", "release_year", "time_range"])
    .copy()
)

# Keep windows in chronological scope order.
window_order = ["Last 4 Weeks", "Last 6 Months", "All Time"]
df_tracks["time_range"] = pd.Categorical(df_tracks["time_range"], categories=window_order, ordered=True)

# Convert year to int to avoid float year labels from null-influenced import.
df_tracks["release_year"] = df_tracks["release_year"].astype(int)
df_tracks["decade"] = (df_tracks["release_year"] // 10) * 10

# Summary table by listening window.
year_summary = (
    df_tracks.groupby("time_range", as_index=False)
    .agg(
        tracks_count=("name", "count"),
        avg_release_year=("release_year", "mean"),
        oldest_release_year=("release_year", "min"),
        newest_release_year=("release_year", "max"),
    )
    .sort_values("time_range")
)
year_summary["avg_release_year"] = year_summary["avg_release_year"].round(0).astype(int)

# Counts by exact year and by decade (used for Q2 visualizations).
year_distribution = (
    df_tracks.pivot_table(
        index="release_year",
        columns="time_range",
        values="name",
        aggfunc="count",
        fill_value=0,
    )
    .reindex(columns=window_order)
    .sort_index()
)

decade_distribution = (
    df_tracks.pivot_table(
        index="decade",
        columns="time_range",
        values="name",
        aggfunc="count",
        fill_value=0,
    )
    .reindex(columns=window_order)
    .sort_index()
)

decade_long = (
    df_tracks.groupby(["time_range", "decade"], as_index=False)
    .size()
    .rename(columns={"size": "tracks_count"})
    .sort_values(["decade", "time_range"])
)

print("Release-year summary by listening window")
display(year_summary)

print("\nTrack count by release year and window")
display(year_distribution)

print("\nTrack count by decade and window")
display(decade_distribution)


Release-year summary by listening window


,time_range,tracks_count,avg_release_year,oldest_release_year,newest_release_year
0,Last 4 Weeks,10,2020,1975,2026
1,Last 6 Months,10,2019,1975,2026
2,All Time,10,2005,1975,2025



Track count by release year and window


time_range,Last 4 Weeks,Last 6 Months,All Time
release_year,,,
1975,1,1,1
1979,0,0,2
1998,0,0,1
2002,0,0,1
2017,1,0,0
2019,0,1,0
2020,0,1,0
2024,0,0,4
2025,0,4,1



Track count by decade and window


time_range,Last 4 Weeks,Last 6 Months,All Time
decade,,,
1970,1,1,3
1990,0,0,1
2000,0,0,1
2010,1,1,0
2020,8,8,5


**Interpretation:**  
The release-year tables indicate that my top tracks lean newer overall, with the strongest concentration in recent years/decades and a smaller tail of older catalog titles. `year_summary` and `year_distribution` show where each listening window clusters by year, while `decade_distribution` confirms that recent decades account for most tracks. This answers Q2: I gravitate more toward newer releases, but still include some older tracks across windows.

**Question 3:** *From my 50 most recently played tracks, what proportion of listens come from albums vs playlist vs other contexts, which artists do I listen to most in album mode vs playlist mode?*

In [9]:

# Analyze 50 recent plays: context breakdown and top artists by context.

from IPython.display import display

required_cols = {"source", "track", "artist", "played_at", "context", "release_year"}
if "df" not in globals() or not required_cols.issubset(df.columns):
    raise NameError("Run the dataset loading cell first so df is available.")

# Pull only recently played records from the exported Spotify dataset.
df_recent = (
    df.loc[df["source"] == "recently_played", ["track", "artist", "played_at", "context", "release_year"]]
    # Exclude incomplete listening events missing core identifiers/timestamp.
    .dropna(subset=["track", "artist", "played_at"])
    .copy()
)

# Normalize timestamps for reliable time ordering and time-based analysis.
df_recent["played_at"] = pd.to_datetime(df_recent["played_at"], errors="coerce")
# Label missing context values explicitly so they are counted as "unknown".
df_recent["context"] = df_recent["context"].fillna("unknown")
df_recent = df_recent.sort_values("played_at", ascending=False)

# Table 1: listening context breakdown.
context_summary = (
    df_recent["context"]
    .value_counts(dropna=False)
    .rename_axis("context")
    .reset_index(name="plays")
)
context_summary["percent"] = (context_summary["plays"] / len(df_recent) * 100).round(1)

# Table 2: top artists overall in recent plays.
top_artists_recent = (
    df_recent["artist"]
    .value_counts()
    .head(10)
    .rename_axis("artist")
    .reset_index(name="plays")
)

# Table 3: top artists within album vs playlist context.
focus_contexts = ["album", "playlist"]
artist_by_context = (
    df_recent[df_recent["context"].isin(focus_contexts)]
    .groupby(["context", "artist"], as_index=False)
    .size()
    .rename(columns={"size": "plays"})
    .sort_values(["context", "plays"], ascending=[True, False])
)
artist_by_context = artist_by_context.groupby("context").head(5)

# Quick recency check (most recent and oldest track in this sample).
most_recent = df_recent.iloc[0]
oldest = df_recent.iloc[-1]

print("Listening context breakdown (recent plays)")
display(context_summary)

print("\nTop artists in recent plays")
display(top_artists_recent)

print("\nTop artists by listening context (album vs playlist)")
display(artist_by_context)

print(
    f"\nMost recent listen: {most_recent['track']} by {most_recent['artist']}\n"
    f"Oldest in sample: {oldest['track']} by {oldest['artist']}"
)


Listening context breakdown (recent plays)


,context,plays,percent
0,playlist,24,48.0
1,album,14,28.0
2,unknown,12,24.0



Top artists in recent plays


,artist,plays
0,Daft Punk,6
1,The Strokes,6
2,Joy Rhodes,2
3,Feid,2
4,Central Cee,2
5,Pale Jay,2
6,Bombay Bicycle Club,1
7,Daryl Hall & John Oates,1
8,Gilsons,1
9,HEAVY CHEST,1



Top artists by listening context (album vs playlist)


,context,artist,plays
4,album,The Strokes,6
1,album,Daft Punk,5
0,album,Bombay Bicycle Club,1
2,album,Daryl Hall & John Oates,1
3,album,Gilsons,1
5,playlist,Bird and Byron,1
6,playlist,Cloak Bay,1
7,playlist,Daft Punk,1
8,playlist,Dal,1
9,playlist,Faye Meana,1



Most recent listen: The Game of Love by Daft Punk
Oldest in sample: Lose Yourself to Dance (feat. Pharrell Williams) by Daft Punk


**Interpretation:**  
`context_summary` shows the proportion of recent listening by context (album, playlist, and unknown which I can safely assume is the DJ feature), and `artist_by_context` shows that the top artists are not identical across album and playlist listening. In other words, where I listen (album vs playlist) changes which artists dominate the recent sample. This answers the question by linking both the context share and the artist-by-context breakdown to my recent listening behavior. Telling me that I explore new artist by playlists and listen to known artist by album.

---

## Section to show my process to this point

To reach the final outputs, I had to iterate the scripts several times because my first run did not give a reliable picture of my listening patterns. The early tables were affected by missing values across endpoint-specific columns, so some summaries looked incomplete or inconsistent. I then iterated each analysis block by adding explicit cleaning steps (`dropna` for required fields, type conversion for numeric rank/year fields, datetime parsing for `played_at`, and filling missing `context` values with `unknown`). After those adjustments, the results became much clearer and better reflected my actual listening habits across artists, release years, and listening contexts.

---

## Section 4 — Visualization

Create at least one visualization that supports one of your analysis findings. Your chart should:

- Have a title that states the finding, not just the data (e.g., "Satisfaction scores drop sharply after age 40" not "Satisfaction by age")
- Have labeled axes
- Use a chart type appropriate for your data (bar for categorical comparison, scatter for relationships, line for trends over time)

Below the chart, explain in a markdown cell: why you chose this chart type, and what you want the reader to take away from it.

# Question 1

In [10]:
# Artist rank movement across listening windows (connected dot plot)
import plotly.express as px

if "artists_rank_df" not in globals():
    raise NameError("Run the Question 1 analysis cell first to create artists_rank_df.")

focus_artists = (
    artists_rank_df.groupby("artist")["rank"]
    .min()
    .sort_values()
    .head(10)
    .index
)

plot_df = (
    artists_rank_df[artists_rank_df["artist"].isin(focus_artists)]
    .dropna(subset=["time_range_key", "rank"])
    .sort_values(["artist", "time_range_key"])
)

# High-contrast, colorblind-friendly palette for clearer differentiation.
artist_palette = [
    "#0072B2", "#E69F00", "#009E73", "#D55E00", "#CC79A7",
    "#56B4E9", "#F0E442", "#000000", "#8A2BE2", "#A6761D",
]

fig = px.line(
    plot_df,
    x="time_range_key",
    y="rank",
    color="artist",
    markers=True,
    color_discrete_sequence=artist_palette,
    category_orders={"time_range_key": ["short_term", "medium_term", "long_term"]},
    labels={
        "time_range_key": "Listening Window",
        "rank": "Artist Rank (1 = highest)",
        "artist": "Artist",
    },
    title="Top artist show clear consistency",
)

fig.update_yaxes(autorange="reversed", dtick=1)
fig.update_traces(line=dict(width=3), marker=dict(size=9))

# Put each artist in its own legend group so extra group gap adds spacing.
fig.for_each_trace(lambda trace: trace.update(legendgroup=trace.name))

fig.update_layout(
    template="plotly_white",
    legend=dict(
        title="Artists",
        orientation="v",
        yanchor="top",
        y=1,
        xanchor="left",
        x=1.02,
        font=dict(size=10),
        tracegroupgap=8,
    ),
    margin=dict(r=200),
)
fig.show()


**Chart rationale:**  
A connected dot plot is a better fit for this question because the main goal is to show how each artist's rank moves across ordered time windows. The connecting lines make changes in direction easy to see while still keeping exact rank values visible at each point. Reversing the y-axis keeps rank #1 at the top, matching how rankings are read. The key takeaway is that some artists remain consistently near the top while others change position as the listening window broadens.


# Question 2

In [11]:
# Q2 Visualization 1: release year strip plot by listening window
if "df_tracks" not in globals() or "window_order" not in globals():
    raise NameError("Run the Question 2 analysis cell first to create df_tracks.")

strip_fig = px.strip(
    df_tracks.sort_values("time_range"),
    x="time_range",
    y="release_year",
    color="time_range",
    category_orders={"time_range": window_order},
    labels={
        "time_range": "Listening Window",
        "release_year": "Release Year",
    },
    title=" In the short term release years are more recent, but older tracks still appear",
)

# Slightly larger markers to improve readability for small samples.
strip_fig.update_traces(marker=dict(size=11, opacity=0.8), jitter=0.35)

# Add mean release-year marker for each window.
means = df_tracks.groupby("time_range", observed=False)["release_year"].mean().reindex(window_order)
strip_fig.add_scatter(
    x=window_order,
    y=means.values,
    mode="markers+lines",
    marker=dict(symbol="diamond", size=12, color="black"),
    line=dict(color="black", dash="dot"),
    name="Average release year",
)

strip_fig.update_layout(template="plotly_white", showlegend=True)
strip_fig.show()

**Chart rationale:**  
A strip plot is the best fit here because it shows every top track as an individual point, making it easy to compare the spread and concentration of release years across listening windows. This view preserves detail that summary tables can hide, including outliers and overlap between windows. The main takeaway is that the short-term window clusters more heavily in recent years, while older tracks still appear across the broader windows.

In [12]:
# Q2 Visualization 2: grouped bar chart by decade and listening window
if "decade_long" not in globals() or "window_order" not in globals():
    raise NameError("Run the Question 2 analysis cell first to create decade_long.")

bars_fig = px.bar(
    decade_long,
    x="decade",
    y="tracks_count",
    color="time_range",
    barmode="group",
    category_orders={"time_range": window_order},
    labels={
        "decade": "Release Decade",
        "tracks_count": "Number of Top Tracks",
        "time_range": "Listening Window",
    },
    title="Recent releases appear prominently in the short term",
)

bars_fig.update_layout(template="plotly_white")
bars_fig.show()

**Chart rationale:**  
A grouped bar chart works well for decade-level comparison because it puts listening windows side by side within each decade, so differences in volume are easy to read. Compared with the strip plot, this chart emphasizes broader era patterns instead of individual tracks. The key takeaway is that recent decades account for most top tracks, with the strongest concentration in the shorter listening window.

# Question 3


In [13]:
# Q3 Visualization 1: donut chart of recent listening context share
import plotly.express as px

if "context_summary" not in globals():
    raise NameError("Run the Question 3 analysis cell first to create context_summary.")

q3_donut = px.pie(
    context_summary,
    names="context",
    values="plays",
    hole=0.55,
    color="context",
    labels={"context": "Listening Context", "plays": "Plays"},
    title="Recent listening distribution show playlist as the most popular",
)

q3_donut.update_traces(
    textposition="inside",
    texttemplate="%{label}<br>%{percent}",
    hovertemplate="Context: %{label}<br>Plays: %{value}<br>Share: %{percent}<extra></extra>",
)
q3_donut.update_layout(template="plotly_white")
q3_donut.show()

**Chart rationale:**  
A donut chart is appropriate for this question because it clearly communicates part-to-whole composition of recent listens by context. The chart makes relative share immediately visible while still showing exact play counts in hover details. The takeaway is that one context category contributes the largest share of recent listening, with album and unknown(Spotify DJ) contexts making up the remainder.

In [14]:
# Q3 Visualization 2: top artists by listening mode (album vs playlist)
import plotly.express as px

if "artist_by_context" not in globals():
    raise NameError("Run the Question 3 analysis cell first to create artist_by_context.")

context_order = ["album", "playlist"]
artist_context_plot = artist_by_context.copy()
artist_context_plot["context"] = pd.Categorical(
    artist_context_plot["context"], categories=context_order, ordered=True
)
artist_context_plot = artist_context_plot.sort_values(["context", "plays"], ascending=[True, True])

q3_context_bars = px.bar(
    artist_context_plot,
    x="plays",
    y="artist",
    color="context",
    orientation="h",
    facet_col="context",
    facet_col_spacing=0.12,
    category_orders={"context": context_order},
    labels={"plays": "Recent Plays", "artist": "Artist", "context": "Listening Context"},
    title="Recent artist listenend to compared by format")

q3_context_bars.update_layout(template="plotly_white", showlegend=False)
q3_context_bars.for_each_annotation(lambda a: a.update(text=a.text.replace("context=", "Context: ")))
q3_context_bars.show()

**Chart rationale:**  
A horizontal grouped bar chart is a strong fit for comparing top artists by listening context because it makes rank and magnitude differences easy to scan within and across contexts. Splitting by context highlights that the artist mix is not identical between album and playlist listening. The key takeaway is that listening mode shapes which artists appear most in recent plays.

In [16]:
# Export Plotly visualizations (Q1–Q3) as PNG using Kaleido
from pathlib import Path
import importlib
import subprocess
import sys

# Ensure kaleido is available in this kernel.
try:
    importlib.import_module("kaleido")
except ImportError:
    print("Installing kaleido...")
    try:
        subprocess.check_call([sys.executable, "-m", "pip", "install", "kaleido", "-q"])
    except subprocess.CalledProcessError:
        # Fallback for externally managed Python environments.
        subprocess.check_call([
            sys.executable,
            "-m",
            "pip",
            "install",
            "kaleido",
            "-q",
            "--break-system-packages",
        ])

output_dir = Path("MP1 Visualizations")
output_dir.mkdir(exist_ok=True)

try:
    _ipy = get_ipython()
except NameError:
    _ipy = None
_ns = _ipy.user_ns if _ipy is not None else globals()

figures_to_save = [
    ("fig", "q1_artist_rank_movement.png"),
    ("strip_fig", "q2_release_year_strip.png"),
    ("bars_fig", "q2_decade_grouped_bar.png"),
    ("q3_donut", "q3_context_donut.png"),
    ("q3_context_bars", "q3_context_top_artists.png"),
]

missing = [name for name, _ in figures_to_save if name not in _ns]
if missing:
    raise NameError(
        "Missing figure objects: " + ", ".join(missing) +
        ". Run all visualization cells (Q1–Q3) before this export cell."
    )

saved_files = []
for fig_name, filename in figures_to_save:
    out_path = output_dir / filename
    _ns[fig_name].write_image(str(out_path), format="png", scale=2)
    saved_files.append(out_path)

print("Saved PNG files:")
for p in saved_files:
    print(f"- {p}")



Saved PNG files:
- MP1 Visualizations/q1_artist_rank_movement.png
- MP1 Visualizations/q2_release_year_strip.png
- MP1 Visualizations/q2_decade_grouped_bar.png
- MP1 Visualizations/q3_context_donut.png
- MP1 Visualizations/q3_context_top_artists.png


---

## Section 5 — Conclusions

Write 3–5 sentences summarizing what you found. Address these questions:

- What is the most important thing your analysis revealed?
- What surprised you?
- What would you investigate next if you had more time or data?
- What are the limitations of this analysis — what can't you conclude from this data?

Then complete the competency claim below.

**Summary of findings:**  

The most important result from this analysis is that my listening behavior combines stability and change across short, medium and long term timeframes, I keep a consistent core of favorite artist but also rotate in newer artist exploration in the short term. This illustrates that my top tracks skew towards recent releases while still including an older catalog in there. What surprised me the most is that these patterns coexist clearly and that playlist make up a large share of recent listening while also showing a more diverse artist mix compared to album listening, suggesting playlist is where I tend to do more exploration. If I had more time I would love to extend the sample size to include a larger number of artist and tracks to analyze from. The main limitation I encountered was the depracted endpoints which would have allowed me to generate more interesting insights as well as the limited number of data that can be extracted per API request. 

---

## Competency Claim

In a `mp1.md` file in your GitHub repository, write a short competency claim (2–4 sentences) for each domain you feel this project demonstrates. Be specific — cite something you actually did in this notebook.

Domains covered by this project typically include:
- **C3 — Data cleaning and file handling** (if you cleaned or reshaped data)
- **C5 — Data analysis with pandas** (answering questions with code)
- **C6 — Data visualization** (your chart)
- **C7 — Critical evaluation and professional judgment** (your interpretation and limitations section)

You don't have to claim every domain — only the ones your work actually demonstrates.